rfcs/0002-mvp.md

```diff
@@ -0,0 +1,159 @@
+│   ├── public.func1.sql                
+│   ├── public.func2.sql│   
+│   # Postgres roles and grants.
+├── roles                                
+│   ├── role_name.sql    
+│   # Types: custom database types
+└── types
+    ├── _init.sql
+    ├── public.some_type.sql
+    └── extensions.some_type.sql
+```
+
+
+### Branching
+
+This is going to be a key part of Supabase development. Since so many of our users are on Vercel/Netlify, we should have the ability to support [branch/preview-development](https://vercel.com/docs/platform/deployments).
+
+We can do this by connecting to a different database, an idea that I got from [pgsh](https://github.com/sastraxi/pgsh). For example:
+
+- `postgresql://postgres@localhost:5432/master`
+- `postgresql://postgres@localhost:5432/develop`
+- `postgresql://postgres@localhost:5432/chore_add_types`
+
+This is the same from CI (like github actions), except that instead of running against a local database, it is running against the supabase database. We will need a way inside our config to match branch names to database names too. For example 
+
+
+### Dump
+
+```sh
+supa db dump                # dumps the currently active (local) database to disk
+supa db restore             # restores the (local) database from disk
+supa migrate                # generate the migrations file
+supa up                     # runs the "up" migrations on Prod
+supa down                   # runs the "down" migrations on Prod
+```
+
+The `supa migrate` task is the trickiest part here. See [#migration-questions](#migration-questions) below.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Supabase is about "time to value", and making things as simple/easy as possible. A CLI can be complicated, so we should make sure that the experience is as amazing as our hosted platform
+
+# Alternatives
+[alternatives]: #alternatives
+
+The hosted platform provides much of this functionality. We could continue to build this functionality into the platform however it will be a great dev experience if they can develop locally and push their changes later.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
+### Migration questions
+[migration-questions]: #migration-questions
+
+The `supa migrate` task is the trickiest part. How do we determine what migrations should run on our database, since the database is created declaratively? We would need to compare the databases somehow, and figure out the `up/down` migrations. Alternatively we could do "stateless" migrations, where we just run `restore` on a brand new database, then re-route traffic, but then how do we get the user's prod data into the new database. Bear in mind that enterprise users might have terrabytes of data, and so copying it across will be slow. 
+
+
+### Other questions
+- what is the prefix? should it be `supabase` (clearer) or `supa` (might have a few people calling it "supper/suppa")
+- Which CLI tool/framework/libraries should we use?
```

Comment:
If we use golang (static binaries, easy to distribute), https://github.com/spf13/cobra is used by most (all?) of the nicer modern CLIs (e.g. kubectl, I think also flyctl under the wraps)

Reply:
Looks great, although I suspect we will need to use `postgres-meta`/`pg-api` for a lot of this functionality. 

Unfortunately it creates a dependency on NodeJS

Reply:
https://github.com/vadimdemedes/ink looks good, apparently Prisma used to use it but I couldn't figure out why they migrated away from it.

Reply:
@soedirgo It had issues on Windows for us (Prisma) and we simplified things a lot so it was not needed anymore.